File: jcv-core/src/main/java/com/ekino/oss/jcv/core/validator/OneParameterTemplateValidator.java
/*
 * Copyright (c) 2019 ekino (https://www.ekino.com/)
 */
package com.ekino.oss.jcv.core.validator;

import com.ekino.oss.jcv.core.initializer.OneParameterComparatorInitializer;
import org.skyscreamer.jsonassert.ValueMatcher;

class OneParameterTemplateValidator<T> extends ValueParameterizedTemplateValidator<T> {

    private final boolean parameterRequired;
    private final OneParameterComparatorInitializer<T> comparatorProvider;

    public OneParameterTemplateValidator(String validatorId, boolean parameterRequired, OneParameterComparatorInitializer<T> comparatorProvider) {
        super(validatorId);
        this.parameterRequired = parameterRequired;
        this.comparatorProvider = comparatorProvider;
    }

    @Override
    protected ValueMatcher<T> getValueComparator(ValidatorTemplateManager validatorTemplateManager) {
        String parameter = validatorTemplateManager.extractParameter(0).orElse(null);

        if (parameterRequired && parameter == null) {
            throw new IllegalArgumentException("No parameter found in validator '" + validatorTemplateManager.extractTemplateContent() + "'");
        }

        return comparatorProvider.initComparator(parameter);
    }
}

File: jcv-core/src/main/kotlin/com/ekino/oss/jcv/core/validator/OneParameterTemplateValidator.kt
/*
 * Copyright (c) 2019 ekino (https://www.ekino.com/)
 */
package com.ekino.oss.jcv.core.validator

import com.ekino.oss.jcv.core.initializer.OneParameterComparatorInitializer
import org.skyscreamer.jsonassert.ValueMatcher

internal class OneParameterTemplateValidator<T>(validatorId: String, private val parameterRequired: Boolean, private val comparatorProvider: OneParameterComparatorInitializer<T>) : ValueParameterizedTemplateValidator<T>(validatorId) {

    override fun getValueComparator(validatorTemplateManager: ValidatorTemplateManager): ValueMatcher<T> {
        val parameter = validatorTemplateManager.extractParameter(0)

        require(!(parameterRequired && parameter == null)) { "No parameter found in validator '" + validatorTemplateManager.extractTemplateContent() + "'" }

        return comparatorProvider.initComparator(parameter)
    }
}
